Flaky test - tailsamplingprocessor - TestConcurrentTraceMapSize

Seen here: https://github.com/open-telemetry/opentelemetry-collector-contrib/runs/5889481381?check_suite_focus=true

```
2022-04-08T18:09:50.3893534Z make -C ./processor/tailsamplingprocessor test
2022-04-08T18:09:50.3912312Z make[2]: Entering directory '/home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/processor/tailsamplingprocessor'
2022-04-08T18:09:50.3965173Z go test -race -v -timeout 300s --tags="" ./...
2022-04-08T18:09:57.2862836Z === RUN   TestCompositeHelper
2022-04-08T18:09:57.2863428Z --- PASS: TestCompositeHelper (0.00s)
2022-04-08T18:09:57.2863715Z === RUN   TestLoadConfig
2022-04-08T18:09:57.2864941Z --- PASS: TestLoadConfig (0.02s)
2022-04-08T18:09:57.2872557Z === RUN   TestCreateDefaultConfig
2022-04-08T18:09:57.2872991Z --- PASS: TestCreateDefaultConfig (0.00s)
2022-04-08T18:09:57.2873298Z === RUN   TestCreateProcessor
2022-04-08T18:09:57.2873635Z --- PASS: TestCreateProcessor (0.02s)
2022-04-08T18:09:57.2873938Z === RUN   TestSequentialTraceArrival
2022-04-08T18:09:57.2874321Z --- PASS: TestSequentialTraceArrival (0.16s)
2022-04-08T18:09:57.2874638Z === RUN   TestConcurrentTraceArrival
2022-04-08T18:09:57.2875007Z --- PASS: TestConcurrentTraceArrival (1.26s)
2022-04-08T18:09:57.2875349Z === RUN   TestSequentialTraceMapSize
2022-04-08T18:09:57.2875722Z --- PASS: TestSequentialTraceMapSize (0.61s)
2022-04-08T18:09:57.2876020Z === RUN   TestConcurrentTraceMapSize
2022-04-08T18:09:57.2876368Z race: limit on 8128 simultaneously alive goroutines is exceeded, dying
2022-04-08T18:09:57.2876957Z FAIL	github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor	5.792s
2022-04-08T18:09:57.2877369Z === RUN   TestBatcherNew
2022-04-08T18:09:57.2877645Z === RUN   TestBatcherNew/invalid_numBatches
2022-04-08T18:09:57.2877959Z === RUN   TestBatcherNew/invalid_batchChannelSize
2022-04-08T18:09:57.2878255Z === RUN   TestBatcherNew/valid
2022-04-08T18:09:57.2878575Z --- PASS: TestBatcherNew (0.00s)
2022-04-08T18:09:57.2879009Z     --- PASS: TestBatcherNew/invalid_numBatches (0.00s)
2022-04-08T18:09:57.2879489Z     --- PASS: TestBatcherNew/invalid_batchChannelSize (0.00s)
2022-04-08T18:09:57.2879908Z     --- PASS: TestBatcherNew/valid (0.00s)
2022-04-08T18:09:57.2880181Z === RUN   TestTypicalConfig
2022-04-08T18:09:57.2880498Z --- PASS: TestTypicalConfig (0.54s)
2022-04-08T18:09:57.2880779Z === RUN   TestMinBufferedChannels
2022-04-08T18:09:57.2881127Z --- PASS: TestMinBufferedChannels (0.59s)
2022-04-08T18:09:57.2881371Z PASS
2022-04-08T18:09:57.2881910Z ok  	github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor/internal/idbatcher	1.157s
2022-04-08T18:09:57.2883529Z === RUN   TestEvaluate_AlwaysSample
2022-04-08T18:09:57.2884160Z --- PASS: TestEvaluate_AlwaysSample (0.00s)
2022-04-08T18:09:57.2884973Z === RUN   TestAndEvaluatorNotSampled
2022-04-08T18:09:57.2885737Z --- PASS: TestAndEvaluatorNotSampled (0.00s)
2022-04-08T18:09:57.2886075Z === RUN   TestAndEvaluatorSampled
2022-04-08T18:09:57.2886885Z --- PASS: TestAndEvaluatorSampled (0.00s)
2022-04-08T18:09:57.2887217Z === RUN   TestCompositeEvaluatorNotSampled
2022-04-08T18:09:57.2887645Z --- PASS: TestCompositeEvaluatorNotSampled (0.00s)
2022-04-08T18:09:57.2887982Z === RUN   TestCompositeEvaluatorSampled
2022-04-08T18:09:57.2888375Z --- PASS: TestCompositeEvaluatorSampled (0.00s)
2022-04-08T18:09:57.2888736Z === RUN   TestCompositeEvaluator_OverflowAlwaysSampled
2022-04-08T18:09:57.2889199Z --- PASS: TestCompositeEvaluator_OverflowAlwaysSampled (0.00s)
2022-04-08T18:09:57.2889583Z === RUN   TestCompositeEvaluatorSampled_AlwaysSampled
2022-04-08T18:09:57.2890043Z --- PASS: TestCompositeEvaluatorSampled_AlwaysSampled (0.00s)
2022-04-08T18:09:57.2890408Z === RUN   TestCompositeEvaluatorThrottling
2022-04-08T18:09:57.2890823Z --- PASS: TestCompositeEvaluatorThrottling (0.00s)
2022-04-08T18:09:57.2891202Z === RUN   TestCompositeEvaluator2SubpolicyThrottling
2022-04-08T18:09:57.2891731Z --- PASS: TestCompositeEvaluator2SubpolicyThrottling (0.00s)
2022-04-08T18:09:57.2892073Z === RUN   TestEvaluate_Latency
2022-04-08T18:09:57.2892398Z === RUN   TestEvaluate_Latency/trace_duration_shorter_than_threshold
2022-04-08T18:09:57.2892774Z === RUN   TestEvaluate_Latency/trace_duration_is_equal_to_threshold
2022-04-08T18:09:57.2893218Z === RUN   TestEvaluate_Latency/total_trace_duration_is_longer_than_threshold_but_every_single_span_is_shorter
2022-04-08T18:09:57.2893654Z --- PASS: TestEvaluate_Latency (0.00s)
2022-04-08T18:09:57.2894181Z     --- PASS: TestEvaluate_Latency/trace_duration_shorter_than_threshold (0.00s)
2022-04-08T18:09:57.2894751Z     --- PASS: TestEvaluate_Latency/trace_duration_is_equal_to_threshold (0.00s)
2022-04-08T18:09:57.2895555Z     --- PASS: TestEvaluate_Latency/total_trace_duration_is_longer_than_threshold_but_every_single_span_is_shorter (0.00s)
2022-04-08T18:09:57.2895947Z === RUN   TestNumericTagFilter
2022-04-08T18:09:57.2896274Z === RUN   TestNumericTagFilter/nonmatching_span_attribute
2022-04-08T18:09:57.2896640Z === RUN   TestNumericTagFilter/span_attribute_with_lower_limit
2022-04-08T18:09:57.2897000Z === RUN   TestNumericTagFilter/span_attribute_with_upper_limit
2022-04-08T18:09:57.2897364Z === RUN   TestNumericTagFilter/span_attribute_below_min_limit
2022-04-08T18:09:57.2897722Z === RUN   TestNumericTagFilter/span_attribute_above_max_limit
2022-04-08T18:09:57.2898102Z --- PASS: TestNumericTagFilter (0.00s)
2022-04-08T18:09:57.2898601Z     --- PASS: TestNumericTagFilter/nonmatching_span_attribute (0.00s)
2022-04-08T18:09:57.2899155Z     --- PASS: TestNumericTagFilter/span_attribute_with_lower_limit (0.00s)
2022-04-08T18:09:57.2899710Z     --- PASS: TestNumericTagFilter/span_attribute_with_upper_limit (0.00s)
2022-04-08T18:09:57.2900401Z     --- PASS: TestNumericTagFilter/span_attribute_below_min_limit (0.00s)
2022-04-08T18:09:57.2900934Z     --- PASS: TestNumericTagFilter/span_attribute_above_max_limit (0.00s)
2022-04-08T18:09:57.2901281Z === RUN   TestProbabilisticSampling
2022-04-08T18:09:57.2901591Z === RUN   TestProbabilisticSampling/100%
2022-04-08T18:09:57.2901900Z === RUN   TestProbabilisticSampling/0%
2022-04-08T18:09:57.2902199Z === RUN   TestProbabilisticSampling/25%
2022-04-08T18:09:57.2902505Z === RUN   TestProbabilisticSampling/33%
2022-04-08T18:09:57.2902899Z === RUN   TestProbabilisticSampling/33%_-_custom_salt
2022-04-08T18:09:57.2903276Z === RUN   TestProbabilisticSampling/-%50
2022-04-08T18:09:57.2903579Z === RUN   TestProbabilisticSampling/150%
2022-04-08T18:09:57.2903957Z --- PASS: TestProbabilisticSampling (3.76s)
2022-04-08T18:09:57.2904418Z     --- PASS: TestProbabilisticSampling/100% (0.53s)
2022-04-08T18:09:57.2904865Z     --- PASS: TestProbabilisticSampling/0% (0.58s)
2022-04-08T18:09:57.2905322Z     --- PASS: TestProbabilisticSampling/25% (0.55s)
2022-04-08T18:09:57.2905783Z     --- PASS: TestProbabilisticSampling/33% (0.51s)
2022-04-08T18:09:57.2906280Z     --- PASS: TestProbabilisticSampling/33%_-_custom_salt (0.56s)
2022-04-08T18:09:57.2906844Z     --- PASS: TestProbabilisticSampling/-%50 (0.51s)
2022-04-08T18:09:57.2907290Z     --- PASS: TestProbabilisticSampling/150% (0.51s)
2022-04-08T18:09:57.2907578Z === RUN   TestRateLimiter
2022-04-08T18:09:57.2907897Z --- PASS: TestRateLimiter (0.00s)
2022-04-08T18:09:57.2908213Z === RUN   TestNewStatusCodeFilter_errorHandling
2022-04-08T18:09:57.2908632Z --- PASS: TestNewStatusCodeFilter_errorHandling (0.00s)
2022-04-08T18:09:57.2908948Z === RUN   TestStatusCodeSampling
2022-04-08T18:09:57.2909379Z === RUN   TestStatusCodeSampling/filter_on_ERROR_-_none_match
2022-04-08T18:09:57.2909824Z === RUN   TestStatusCodeSampling/filter_on_OK_and_ERROR_-_none_match
2022-04-08T18:09:57.2910261Z === RUN   TestStatusCodeSampling/filter_on_UNSET_-_matches
2022-04-08T18:09:57.2910685Z === RUN   TestStatusCodeSampling/filter_on_OK_and_UNSET_-_matches
2022-04-08T18:09:57.2911069Z --- PASS: TestStatusCodeSampling (0.00s)
2022-04-08T18:09:57.2911591Z     --- PASS: TestStatusCodeSampling/filter_on_ERROR_-_none_match (0.00s)
2022-04-08T18:09:57.2912156Z     --- PASS: TestStatusCodeSampling/filter_on_OK_and_ERROR_-_none_match (0.00s)
2022-04-08T18:09:57.2912699Z     --- PASS: TestStatusCodeSampling/filter_on_UNSET_-_matches (0.00s)
2022-04-08T18:09:57.2913250Z     --- PASS: TestStatusCodeSampling/filter_on_OK_and_UNSET_-_matches (0.00s)
2022-04-08T18:09:57.2913576Z === RUN   TestStringTagFilter
2022-04-08T18:09:57.2913899Z === RUN   TestStringTagFilter/nonmatching_node_attribute_key
2022-04-08T18:09:57.2914268Z === RUN   TestStringTagFilter/nonmatching_node_attribute_value
2022-04-08T18:09:57.2914616Z === RUN   TestStringTagFilter/matching_node_attribute
2022-04-08T18:09:57.2914966Z === RUN   TestStringTagFilter/nonmatching_span_attribute_key
2022-04-08T18:09:57.2915389Z === RUN   TestStringTagFilter/nonmatching_span_attribute_value
2022-04-08T18:09:57.2915742Z === RUN   TestStringTagFilter/matching_span_attribute
2022-04-08T18:09:57.2916094Z === RUN   TestStringTagFilter/matching_span_attribute_with_regex
2022-04-08T18:09:57.2916476Z === RUN   TestStringTagFilter/nonmatching_span_attribute_with_regex
2022-04-08T18:09:57.2916916Z === RUN   TestStringTagFilter/matching_span_attribute_with_regex_without_CacheSize_provided_in_config
2022-04-08T18:09:57.2917350Z === RUN   TestStringTagFilter/matching_plain_text_node_attribute_in_regex
2022-04-08T18:09:57.2917757Z === RUN   TestStringTagFilter/nonmatching_span_attribute_on_empty_filter_list
2022-04-08T18:09:57.2918150Z === RUN   TestStringTagFilter/invert_nonmatching_node_attribute_key
2022-04-08T18:09:57.2918534Z === RUN   TestStringTagFilter/invert_nonmatching_node_attribute_value
2022-04-08T18:09:57.2918912Z === RUN   TestStringTagFilter/invert_nonmatching_node_attribute_list
2022-04-08T18:09:57.2919286Z === RUN   TestStringTagFilter/invert_matching_node_attribute
2022-04-08T18:09:57.2919646Z === RUN   TestStringTagFilter/invert_matching_node_attribute_list
2022-04-08T18:09:57.2920020Z === RUN   TestStringTagFilter/invert_nonmatching_span_attribute_key
2022-04-08T18:09:57.2920406Z === RUN   TestStringTagFilter/invert_nonmatching_span_attribute_value
2022-04-08T18:09:57.2920786Z === RUN   TestStringTagFilter/invert_nonmatching_span_attribute_list
2022-04-08T18:09:57.2921157Z === RUN   TestStringTagFilter/invert_matching_span_attribute
2022-04-08T18:09:57.2921515Z === RUN   TestStringTagFilter/invert_matching_span_attribute_list
2022-04-08T18:09:57.2921888Z === RUN   TestStringTagFilter/invert_matching_span_attribute_with_regex
2022-04-08T18:09:57.2922286Z === RUN   TestStringTagFilter/invert_matching_span_attribute_with_regex_list
2022-04-08T18:09:57.2922688Z === RUN   TestStringTagFilter/invert_nonmatching_span_attribute_with_regex
2022-04-08T18:09:57.2923101Z === RUN   TestStringTagFilter/invert_nonmatching_span_attribute_with_regex_list
2022-04-08T18:09:57.2923515Z === RUN   TestStringTagFilter/invert_matching_plain_text_node_attribute_in_regex
2022-04-08T18:09:57.2923936Z === RUN   TestStringTagFilter/invert_matching_plain_text_node_attribute_in_regex_list
2022-04-08T18:09:57.2924420Z === RUN   TestStringTagFilter/invert_nonmatching_span_attribute_on_empty_filter_list
2022-04-08T18:09:57.2924832Z --- PASS: TestStringTagFilter (0.01s)
2022-04-08T18:09:57.2925324Z     --- PASS: TestStringTagFilter/nonmatching_node_attribute_key (0.00s)
2022-04-08T18:09:57.2925873Z     --- PASS: TestStringTagFilter/nonmatching_node_attribute_value (0.00s)
2022-04-08T18:09:57.2926392Z     --- PASS: TestStringTagFilter/matching_node_attribute (0.00s)
2022-04-08T18:09:57.2926913Z     --- PASS: TestStringTagFilter/nonmatching_span_attribute_key (0.00s)
2022-04-08T18:09:57.2927451Z     --- PASS: TestStringTagFilter/nonmatching_span_attribute_value (0.00s)
2022-04-08T18:09:57.2927963Z     --- PASS: TestStringTagFilter/matching_span_attribute (0.00s)
2022-04-08T18:09:57.2928493Z     --- PASS: TestStringTagFilter/matching_span_attribute_with_regex (0.00s)
2022-04-08T18:09:57.2929062Z     --- PASS: TestStringTagFilter/nonmatching_span_attribute_with_regex (0.00s)
2022-04-08T18:09:57.2929734Z     --- PASS: TestStringTagFilter/matching_span_attribute_with_regex_without_CacheSize_provided_in_config (0.00s)
2022-04-08T18:09:57.2930355Z     --- PASS: TestStringTagFilter/matching_plain_text_node_attribute_in_regex (0.00s)
2022-04-08T18:09:57.2930964Z     --- PASS: TestStringTagFilter/nonmatching_span_attribute_on_empty_filter_list (0.00s)
2022-04-08T18:09:57.2931540Z     --- PASS: TestStringTagFilter/invert_nonmatching_node_attribute_key (0.00s)
2022-04-08T18:09:57.2932117Z     --- PASS: TestStringTagFilter/invert_nonmatching_node_attribute_value (0.00s)
2022-04-08T18:09:57.2932694Z     --- PASS: TestStringTagFilter/invert_nonmatching_node_attribute_list (0.00s)
2022-04-08T18:09:57.2933229Z     --- PASS: TestStringTagFilter/invert_matching_node_attribute (0.00s)
2022-04-08T18:09:57.2933824Z     --- PASS: TestStringTagFilter/invert_matching_node_attribute_list (0.00s)
2022-04-08T18:09:57.2934391Z     --- PASS: TestStringTagFilter/invert_nonmatching_span_attribute_key (0.00s)
2022-04-08T18:09:57.2934972Z     --- PASS: TestStringTagFilter/invert_nonmatching_span_attribute_value (0.00s)
2022-04-08T18:09:57.2935549Z     --- PASS: TestStringTagFilter/invert_nonmatching_span_attribute_list (0.00s)
2022-04-08T18:09:57.2936089Z     --- PASS: TestStringTagFilter/invert_matching_span_attribute (0.00s)
2022-04-08T18:09:57.2936635Z     --- PASS: TestStringTagFilter/invert_matching_span_attribute_list (0.00s)
2022-04-08T18:09:57.2937208Z     --- PASS: TestStringTagFilter/invert_matching_span_attribute_with_regex (0.00s)
2022-04-08T18:09:57.2937807Z     --- PASS: TestStringTagFilter/invert_matching_span_attribute_with_regex_list (0.00s)
2022-04-08T18:09:57.2938415Z     --- PASS: TestStringTagFilter/invert_nonmatching_span_attribute_with_regex (0.00s)
2022-04-08T18:09:57.2939033Z     --- PASS: TestStringTagFilter/invert_nonmatching_span_attribute_with_regex_list (0.00s)
2022-04-08T18:09:57.2939658Z     --- PASS: TestStringTagFilter/invert_matching_plain_text_node_attribute_in_regex (0.00s)
2022-04-08T18:09:57.2940438Z     --- PASS: TestStringTagFilter/invert_matching_plain_text_node_attribute_in_regex_list (0.00s)
2022-04-08T18:09:57.2941084Z     --- PASS: TestStringTagFilter/invert_nonmatching_span_attribute_on_empty_filter_list (0.00s)
2022-04-08T18:09:57.2941448Z === RUN   TestTimeProvider
2022-04-08T18:09:57.2941766Z --- PASS: TestTimeProvider (0.00s)
2022-04-08T18:09:57.2942003Z PASS
2022-04-08T18:09:57.2942527Z ok  	github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor/internal/sampling	3.809s
2022-04-08T18:09:57.2942901Z FAIL
2022-04-08T18:09:57.3029110Z make[2]: *** [../../Makefile.Common:46: test] Error 1
2022-04-08T18:09:57.3029865Z make[2]: Leaving directory '/home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib/processor/tailsamplingprocessor'
2022-04-08T18:09:57.3031449Z make[1]: *** [Makefile:163: for-all-target-./processor/tailsamplingprocessor] Error 2
2022-04-08T18:09:57.3033096Z make: *** [Makefile:73: gotest] Error 2
2022-04-08T18:09:57.3035683Z make[1]: Leaving directory '/home/runner/work/opentelemetry-collector-contrib/opentelemetry-collector-contrib'
2022-04-08T18:09:57.3089275Z ##[error]Process completed with exit code 2.
```

@gouthamve's other flaky hunting reminded me of goleak, and I used it to try to figure out what's going on here. Not sure yet this is relevant, but there's indeed a leak of two goroutines:

```
=== RUN   TestConcurrentTraceMapSize
    leaks.go:78: found unexpected goroutines:
        [Goroutine 6 in state select, with go.opencensus.io/stats/view.(*worker).start on top of the stack:
        goroutine 6 [select]:
        go.opencensus.io/stats/view.(*worker).start(0xc00003a080)
        	/home/jpkroehling/go/pkg/mod/go.opencensus.io@v0.23.0/stats/view/worker.go:276 +0x177
        created by go.opencensus.io/stats/view.init.0
        	/home/jpkroehling/go/pkg/mod/go.opencensus.io@v0.23.0/stats/view/worker.go:34 +0xf7
        
         Goroutine 11 in state chan receive, with github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor/internal/idbatcher.New.func1 on top of the stack:
        goroutine 11 [chan receive]:
        github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor/internal/idbatcher.New.func1()
        	/home/jpkroehling/Projects/src/github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor/internal/idbatcher/id_batcher.go:106 +0x79
        created by github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor/internal/idbatcher.New
        	/home/jpkroehling/Projects/src/github.com/open-telemetry/opentelemetry-collector-contrib/processor/tailsamplingprocessor/internal/idbatcher/id_batcher.go:105 +0x3bf
        ]
--- FAIL: TestConcurrentTraceMapSize (1.08s)
```

There's not much we can do for the opencensus one, but the other one indeed is suspicious. I'll keep looking into this.

--- a/processor/tailsamplingprocessor/internal/idbatcher/id_batcher_test.go
+++ b/processor/tailsamplingprocessor/internal/idbatcher/id_batcher_test.go
@@ -103,11 +103,17 @@ func concurrencyTest(t *testing.T, numBatches, newBatchesInitialCapacity, batchC
 
 	ids := generateSequentialIds(10000)
 	wg := &sync.WaitGroup{}
+	// Limit the concurrency here to avoid creating too many goroutines and hit
+	// https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/9126
+	concurrencyLimiter := make(chan struct{}, 128)
+	defer close(concurrencyLimiter)
 	for i := 0; i < len(ids); i++ {
 		wg.Add(1)
+		concurrencyLimiter <- struct{}{}
 		go func(id pcommon.TraceID) {
 			batcher.AddToCurrentBatch(id)
 			wg.Done()
+			<-concurrencyLimiter
 		}(ids[i])
 	}
 

--- a/processor/tailsamplingprocessor/processor_test.go
+++ b/processor/tailsamplingprocessor/processor_test.go
@@ -208,16 +208,24 @@ func TestConcurrentTraceArrival(t *testing.T) {
 		require.NoError(t, tsp.Shutdown(context.Background()))
 	}()
 
+	// Limit the concurrency here to avoid creating too many goroutines and hit
+	// https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/9126
+	concurrencyLimiter := make(chan struct{}, 128)
+	defer close(concurrencyLimiter)
 	for _, batch := range batches {
 		// Add the same traceId twice.
 		wg.Add(2)
+		concurrencyLimiter <- struct{}{}
 		go func(td ptrace.Traces) {
 			require.NoError(t, tsp.ConsumeTraces(context.Background(), td))
 			wg.Done()
+			<-concurrencyLimiter
 		}(batch)
+		concurrencyLimiter <- struct{}{}
 		go func(td ptrace.Traces) {
 			require.NoError(t, tsp.ConsumeTraces(context.Background(), td))
 			wg.Done()
+			<-concurrencyLimiter
 		}(batch)
 	}
 

